wdl_runner/file_util.py

```diff
@@ -9,96 +9,142 @@
 # file_util.py
 
 import logging
-import simplejson
 import string
-import subprocess
+from pathlib import Path
+from typing import List
+from urllib.parse import urlparse
 
+import simplejson
+from google.cloud import storage
 from googleapiclient import discovery
 from googleapiclient.errors import HttpError
 
-import sys_util
+import wdl_runner.sys_util
 
 
 def file_safe_substitute(file_name, mapping):
-  """Performs placeholder replacement on a file, saving contents in place."""
-
-  with open(file_name, 'r') as f:
-    file_contents = f.read()
-    return string.Template(file_contents).safe_substitute(mapping)
-
-
-def gsutil_cp(source_files, dest_dir):
-  """Copies files to GCS and exits on error."""
-
-  cp_cmd = ['gsutil', 'cp'] + source_files + [dest_dir]
-
-  logging.info("Copying %s to %s", source_files, dest_dir)
-
-  # Copies can fail, so include retries...
-  for attempt in range(3):
-    p = subprocess.Popen(cp_cmd, stderr=subprocess.PIPE)
-    return_code = p.wait()
-    if not return_code:
-      return
-
-    logging.warn("Copy %s to %s failed: attempt %d",
-                 source_files, dest_dir, attempt)
-
-  sys_util.exit_with_error(
-      "copying files from %s to %s failed: %s" % (
-          source_files, dest_dir, p.stderr.read()))
+    """Performs placeholder replacement on a file, saving contents in place."""
+
+    with open(file_name, "r") as f:
+        file_contents = f.read()
+        return string.Template(file_contents).safe_substitute(mapping)
+
+
+def _upload_one_file(client, src_file, dest_bucket, dest_gs_url):
+    src_parsed_url = urlparse(src_file)
+    dest_parsed_url = urlparse(dest_gs_url)
+
+    if src_parsed_url.scheme == "":  # local file upload
+
+        def once():
+            blob = dest_bucket.blob(
+                str(
+                    Path(
+                        dest_parsed_url.path[1:],
+                        Path(src_file).name,
+                    )
+                )
+            )
+            blob.upload_from_filename(src_file)
+
+    elif src_parsed_url.scheme == "gs":  # copy between gcs buckets
+
+        def once():
+            src_bucketname = src_parsed_url.hostname
+            src_bucket = client.bucket(src_bucketname)
+            path = src_parsed_url.path
+            src_blob = src_bucket.blob(path[1:])
+            src_bucket.copy_blob(
+                src_blob,
+                dest_bucket,
+                str(dest_parsed_url.path[1:]),
+            )
+
+    else:
+        raise ValueError(f"Cannot handle src file {src_file}")
+
+    for attempt in range(3):
```

Comment:
Might have been better to make this a func that yields.

Reply:
I also like having a `@retry` decorator similar to [what Toil has done](https://github.com/DataBiosphere/toil/blob/a98acdb5cbe0f850b2c11403d147577d9971f4e1/src/toil/lib/retry.py#L228) (there are lot of retries in that tool though!). But that's for another day :)